File: simulation/engine.py
"""Deterministic daily simulation orchestration."""
from simulation import (
    actions, contracts, crop_growth, derived, economy_rules, inventory, markets, processing, weather,
)


def _legacy_run_day(player, agent, crops, crops_by_id, upgrades, upgrades_by_id, watering_settings, fertilizer_config, rng):
    player.slot_days += player.slots_total
    player.occupied_slot_days += len(player.planted)
    watered = actions.water_farm(player, agent, crops_by_id, rng, watering_settings)
    harvested = actions.harvest_mature(player, crops_by_id, rng, watering_settings, fertilizer_config)
    _revenue, sold = actions.sell_all(player, crops_by_id, rng)
    upgrade_bought = False
    for upgrade in upgrades:
        if upgrade["id"] not in player.upgrades_owned and agent.should_buy_upgrade(player, upgrade):
            upgrade_bought = actions.buy_upgrade(player, upgrade) or upgrade_bought
    planted_something = _plant_open_slots(
        player, agent, crops, crops_by_id, upgrades_by_id, fertilizer_config
    )
    _finish_day(player, crops, bool(watered or harvested or sold or upgrade_bought or planted_something))


def _effective_storage(base: dict, player, upgrades_by_id: dict, lookups) -> dict:
    return lookups.effective_storage(base, player.upgrades_owned, upgrades_by_id)


def _processing_capacity(config: dict, player, upgrades_by_id: dict, lookups) -> int:
    return lookups.processing_capacity(config, player.upgrades_owned, upgrades_by_id)


def run_day(player, agent, crops: list, crops_by_id: dict, upgrades: list, upgrades_by_id: dict, *args, world=None) -> None:
    if len(args) == 1:
        watering_settings, fertilizer_config, rng = actions.DEFAULT_WATERING, actions.DEFAULT_FERTILIZER, args[0]
    elif len(args) == 3:
        watering_settings, fertilizer_config, rng = args
        watering_settings = watering_settings or actions.DEFAULT_WATERING
        fertilizer_config = fertilizer_config or actions.DEFAULT_FERTILIZER
    else:
        raise TypeError("run_day expects rng or watering_settings, fertilizer_config, rng")
    if not world:
        _legacy_run_day(
            player, agent, crops, crops_by_id, upgrades, upgrades_by_id,
            watering_settings, fertilizer_config, rng,
        )
        return

    player.slot_days += player.slots_total
    player.occupied_slot_days += len(player.planted)
    # Item/recipe/channel indexes are pure functions of the world config, so
    # they are built once per (world, crops) pair instead of being rebuilt on
    # every simulated day of every run.
    lookups = derived.world_lookups(world, crops_by_id)
    items_by_id = lookups.items_by_id
    watering_config = lookups.watering
    fertilizer = lookups.fertilizer
    player.crop_catalog = crops_by_id
    player.upgrades_catalog = upgrades_by_id
    player.contract_config = lookups.contracts

    player.current_weather = weather.generate_weather(player.day, lookups.weather, rng)
    weather.apply_weather(
        player, crops_by_id, player.current_weather, crop_growth,
        lookups.crop_profiles, lookups.plot_regen,
    )
    harvested = actions.harvest_mature(player, crops_by_id, rng, watering_config, fertilizer)
    storage = _effective_storage(lookups.storage_config, player, upgrades_by_id, lookups)
    spoiled = inventory.age_and_spoil(player, storage)
    completed = processing.complete_jobs(player)
    markets.update_daily_prices(player, items_by_id, lookups.markets, rng, lookups.market_profiles)
    player.market_channels = lookups.channels
    new_offers = contracts.generate_offers(player, lookups.contracts, lookups.buyers, items_by_id, rng)

    acted = bool(harvested or spoiled or completed)
    for contract_id in agent.choose_contracts(player, list(new_offers)):
        acted = contracts.accept(player, contract_id) or acted
    for decision in agent.choose_contract_deliveries(player):
        _revenue, delivered = contracts.deliver(player, decision["contract_id"], decision["quantity"])
        acted = delivered > 0 or acted

    recipes_by_id = lookups.recipes_by_id
    capacity = _processing_capacity(lookups.processing, player, upgrades_by_id, lookups)
    for decision in agent.choose_processing(player, lookups.recipes, items_by_id):
        recipe = recipes_by_id.get(decision["recipe_id"])
        if recipe:
            acted = processing.start_job(player, recipe, decision.get("batches", 1), capacity) or acted

    channels = lookups.channels
    channels_by_id = lookups.channels_by_id
    for decision in agent.choose_sales(player, channels, items_by_id):
        channel = channels_by_id.get(decision["channel_id"])
        if channel:
            _revenue, sold = markets.sell(player, decision["item_id"], decision["quantity"], channel)
            acted = sold > 0 or acted

    for upgrade in upgrades:
        if upgrade["id"] not in player.upgrades_owned and agent.should_buy_upgrade(player, upgrade):
            acted = actions.buy_upgrade(player, upgrade) or acted

    for planted in list(player.planted):
        crop = crops_by_id[planted.crop_id]
        if agent.should_water(player, planted, crop) and rng.chance(agent.watering_diligence):
            acted = actions.water_crop(player, planted, watering_config) or acted
        if agent.should_fertilize(player, planted, crop, fertilizer) and not planted.fertilized:
            if player.fertilizer_inventory == 0:
                actions.buy_fertilizer(player, fertilizer)
            acted = actions.fertilize_crop(player, planted, fertilizer) or acted

    acted = _plant_open_slots(
        player, agent, crops, crops_by_id, upgrades_by_id, fertilizer
    ) or acted
    contracts.resolve_expired(player)
    _finish_day(player, crops, acted)


def _plant_open_slots(player, agent, crops, crops_by_id, upgrades_by_id, fertilizer_config=None) -> bool:
    planted_something = False
    while player.open_slots > 0:
        for candidate in crops:
            unlocked = economy_rules.is_crop_unlocked(candidate, player)
            affordable = player.money >= candidate["seed_cost"]
            player.observe_crop_decision(candidate, unlocked, affordable)
        crop = agent.choose_crop(player, crops, crops_by_id, upgrades_by_id)
        if crop is None:
            break
        unlocked = economy_rules.is_crop_unlocked(crop, player)
        affordable = player.money >= crop["seed_cost"]
        if not unlocked or not affordable:
            player.observe_crop_decision(
                crop,
                unlocked,
                affordable,
                blocked_reason="locked" if not unlocked else "unaffordable",
                count_opportunity=False,
            )
            break
        use_fertilizer = bool(
            fertilizer_config
            and agent.should_use_fertilizer(player, crop, fertilizer_config)
        )
        if use_fertilizer and player.fertilizer_inventory == 0:
            combined_cost = crop["seed_cost"] + fertilizer_config["cost"]
            if player.money >= combined_cost:
                actions.buy_fertilizer(player, fertilizer_config, 1)
            else:
                # Fertilizer is optional for planting; do not buy it if the
                # same cash cannot also cover the seed.
                use_fertilizer = False
        use_fertilizer = use_fertilizer and player.fertilizer_inventory > 0
        if not actions.buy_seeds(player, crop, 1):
            break
        player.observe_crop_decision(crop, True, True, selected=True, count_opportunity=False)
        growth_days = economy_rules.effective_growth_days(crop, player, upgrades_by_id)
        planted_something = actions.plant_seed(
            player, crop, growth_days, fertilized=use_fertilizer, fertilizer_config=fertilizer_config
        ) or planted_something
    return planted_something


def _finish_day(player, crops: list, acted: bool) -> None:
    if not acted:
        actions.do_nothing(player)
    player.lowest_money = min(player.lowest_money, player.money)
    player.highest_money = max(player.highest_money, player.money)
    cheapest_seed = derived.cheapest_seed_cost(crops)
    if (
        player.money < cheapest_seed
        and not player.planted
        and not player.processing_jobs
        and not any(lot.quantity > 0 for lot in player.inventory_lots)
        and not any(player.crop_inventory.values())
    ):
        player.bankrupt = True
        player.bankruptcy_day = player.day + 1
        player.bankruptcy_reason = "no_viable_reinvestment"
    player.day += 1

File: farm-c/src/engine.c
#include "engine.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "actions.h"
#include "contracts.h"
#include "economy.h"
#include "inventory.h"
#include "markets.h"
#include "processing.h"
#include "rng_hash.h"
#include "weather.h"

static void set_error(EngineError *error, EngineErrorCode code,
                      const char *message) {
  if (error == NULL)
    return;
  error->code = code;
  snprintf(error->message, sizeof(error->message), "%s", message);
}

/* The four decision-buffer steps (12/13/14/15 below) all share this shape:
 * call the agent, and if the buffer's own growth allocation failed, free it
 * and bail with a matching error message. */
#define ENGINE_CHECK_ALLOC(buf, free_fn, what)                               \
  do {                                                                       \
    if ((buf).allocation_failed) {                                           \
      free_fn(&(buf));                                                       \
      set_error(error, ENGINE_ERROR_ALLOCATION, what " buffer allocation failed"); \
      return false;                                                          \
    }                                                                        \
  } while (0)

#define ENGINE_CHECK_STATE_ALLOC(what)                                        \
  do {                                                                        \
    if (state->allocation_failed || contracts_had_allocation_failure() ||     \
        rng_hash_had_allocation_failure()) {                                  \
      state->allocation_failed = true;                                        \
      set_error(error, ENGINE_ERROR_ALLOCATION, what " allocation failed"); \
      return false;                                                           \
    }                                                                         \
  } while (0)

static StorageConfig effective_storage(const FarmState *state) {
  StorageConfig storage = state->config->storage;
  for (size_t i = 0; i < state->config->upgrade_count; i++) {
    const UpgradeDef *upgrade = &state->config->upgrades[i];
    if (state->upgrades_owned[upgrade->id] &&
        upgrade->effect.type == EFFECT_STORAGE) {
      storage.capacity += upgrade->effect.as.storage.capacity_bonus;
      storage.shelf_life_multiplier *=
          upgrade->effect.as.storage.shelf_life_multiplier;
    }
  }
  return storage;
}

static bool crop_can_be_funded(const FarmState *state, const CropDef *crop) {
  return state->money >= crop->seed_cost ||
         state->seed_inventory[crop->item_id] > 0;
}

static bool plant_open_slots(FarmState *state, const Agent *agent) {
  bool acted = false;
  int open_slots = farm_state_open_slots(state);
  if (open_slots > 0) {
    if ((size_t)open_slots > SIZE_MAX - state->planted.count ||
        !vec_reserve((void **)&state->planted.data, &state->planted.capacity,
                     state->planted.count + (size_t)open_slots,
                     sizeof(*state->planted.data))) {
      farm_state_mark_allocation_failed(state);
      return false;
    }
  }
  /* next_plot only ever advances across iterations, so filling k of n open
   * slots is a single O(n) sweep instead of an O(n*k) rescan-from-0 every
   * time a slot gets planted. */
  size_t next_plot = 0;
  while (farm_state_open_slots(state) > 0) {
    bool free_plot = false;
    for (; next_plot < state->plot_count; next_plot++) {
      if (state->plots[next_plot].planted_index == -1) {
        free_plot = true;
        break;
      }
    }
    if (!free_plot)
      break;
    ItemId crop_id = agent->choose_crop(agent, state, state->config);
    if (contracts_had_allocation_failure() || rng_hash_had_allocation_failure()) {
      farm_state_mark_allocation_failed(state);
      return false;
    }
    if (!id_valid(crop_id))
      break;
    const CropDef *crop = config_find_crop(state->config, crop_id);
    if (crop == NULL || !economy_is_crop_unlocked(crop, state) ||
        !crop_can_be_funded(state, crop))
      break;

    bool fertilized = agent->should_use_fertilizer(agent, state, crop_id);
    if (contracts_had_allocation_failure() || rng_hash_had_allocation_failure()) {
      farm_state_mark_allocation_failed(state);
      return false;
    }
    if (fertilized && state->fertilizer_inventory == 0) {
      if (state->money >= crop->seed_cost + state->config->fertilizer.cost) {
        (void)actions_buy_fertilizer(state, &state->config->fertilizer, 1);
      } else {
        fertilized = false;
      }
    }
    fertilized = fertilized && state->fertilizer_inventory > 0;
    if (state->seed_inventory[crop_id] <= 0 &&
        !actions_buy_seeds(state, crop, 1))
      break;
    int growth_days = economy_effective_growth_days(crop, state, state->config);
    if (!actions_plant_seed(state, crop, growth_days, fertilized,
                            &state->config->fertilizer))
      break;
    acted = true;
  }
  return acted;
}

static bool no_viable_reinvestment(const FarmState *state) {
  /* Cheap O(1) checks first: on the overwhelming majority of days this
   * returns false right here, so the crop-catalog scan below (invariant
   * for the config's lifetime) only runs on the rare days it's needed. */
  if (state->planted.count != 0 || state->processing_jobs.count != 0)
    return false;
  double cheapest = 0.0;
  bool found = false;
  for (size_t i = 0; i < state->config->crop_count; i++) {
    double cost = state->config->crops[i].seed_cost;
    if (!found || cost < cheapest) {
      cheapest = cost;
      found = true;
    }
  }
  if (!found || state->money >= cheapest)
    return false;
  for (size_t i = 0; i < state->inventory_lots.count; i++) {
    if (state->inventory_lots.data[i].quantity > 0)
      return false;
  }
  for (size_t i = 0; i < state->config->crop_count; i++) {
    if (state->seed_inventory[state->config->crops[i].item_id] > 0)
      return false;
  }
  return true;
}

/*
 *
 */
bool engine_run_day_observed(FarmState *state, const Agent *agent, FarmRng *rng,
                             EngineDayCallback callback, void *context,
                             EngineError *error) {
  if (error != NULL)
    *error = (EngineError){0};
  if (state == NULL || state->config == NULL || agent == NULL || rng == NULL ||
      agent->choose_crop == NULL || agent->should_buy_upgrade == NULL ||
      agent->should_water == NULL || agent->should_fertilize == NULL ||
      agent->choose_contracts == NULL ||
      agent->choose_contract_deliveries == NULL ||
      agent->choose_processing == NULL || agent->choose_sales == NULL ||
      agent->should_use_fertilizer == NULL) {
    set_error(error, ENGINE_ERROR_ARGUMENT,
              "state, config, agent, and rng are required");
    return false;
  }
  contracts_clear_allocation_failure();
  rng_hash_clear_allocation_failure();
  if (state->allocation_failed) {
    set_error(error, ENGINE_ERROR_ALLOCATION, "state allocation previously failed");
    return false;
  }
  const ResolvedConfig *config = state->config;
  state->processing_capacity = config->processing_capacity;
  for (size_t i = 0; i < config->upgrade_count; i++) {
    const UpgradeDef *upgrade = &config->upgrades[i];
    if (state->upgrades_owned[upgrade->id] &&
        upgrade->effect.type == EFFECT_PROCESSING_CAPACITY)
      state->processing_capacity += upgrade->effect.as.processing_capacity;
  }
  /* Computed fresh above every day the full engine runs, so it's always
   * trustworthy from here on -- has_processing_capacity only reads as false
   * for a bare FarmState that never went through run_day (e.g. a direct
   * unit-test construction), matching state.h's contract with contracts.c
   * and profit_optimizer.c's choose_processing. Leaving this unset left
   * both permanently reading processing_capacity as 0, so choose_processing
   * returned early every day of every run -- see docs/agent-decision
   * divergence notes. */
  state->has_processing_capacity = true;
  StorageConfig storage = effective_storage(state);
  bool acted = false;

  /* 1. */ state->slot_days += state->slots_total;
  /* 2. */ state->occupied_slot_days += (int)state->planted.count;
  /* 3. */ WeatherDay weather = weather_generate(state->day, config, rng);
  state->current_season = weather.season;
  /* 4. */ weather_apply(state, &weather);
  /* 5. */ double liability =
      inventory_capture_storage_liability(state, &storage);
  /* 6. */ acted = actions_harvest_mature(state, config, rng, &config->watering,
                                           &config->fertilizer) ||
                    acted;
  ENGINE_CHECK_STATE_ALLOC("harvest");
  /* 7. */ int spoiled = inventory_age_and_spoil(state, &storage, false);
  ENGINE_CHECK_STATE_ALLOC("inventory aging");
  acted = spoiled > 0 || acted;
  /* 8. */ int completed = processing_complete_jobs(state);
  ENGINE_CHECK_STATE_ALLOC("processing completion");
  acted = completed > 0 || acted;
  /* 9. */ spoiled =
      inventory_enforce_storage_capacity(state, storage.capacity);
  ENGINE_CHECK_STATE_ALLOC("storage enforcement");
  acted = spoiled > 0 || acted;
  /* 10. */ markets_update_daily_prices(state, config, rng);
  /* 11. */ contracts_generate_offers(state, config, rng);
  ENGINE_CHECK_STATE_ALLOC("contract offer");

  /* 12. */ ContractDecisionBuffer accepts = {0};
  agent->choose_contracts(agent, state, config, &accepts);
  ENGINE_CHECK_ALLOC(accepts, contract_decision_free, "contract decision");
  ENGINE_CHECK_STATE_ALLOC("contract decision");
  for (size_t i = 0; i < accepts.count; i++) {
    acted = contracts_accept(state, config, accepts.data[i]) || acted;
    if (state->allocation_failed) {
      contract_decision_free(&accepts);
      set_error(error, ENGINE_ERROR_ALLOCATION, "contract acceptance allocation failed");
      return false;
    }
  }
  contract_decision_free(&accepts);
  ENGINE_CHECK_STATE_ALLOC("contract acceptance");
  /* 13. */ DeliveryDecisionBuffer deliveries = {0};
  agent->choose_contract_deliveries(agent, state, &deliveries);
  ENGINE_CHECK_ALLOC(deliveries, delivery_decision_free, "delivery decision");
  ENGINE_CHECK_STATE_ALLOC("delivery decision");
  for (size_t i = 0; i < deliveries.count; i++) {
    int delivered = 0;
    (void)contracts_deliver(state, config, deliveries.data[i].contract_id,
                            deliveries.data[i].quantity, &delivered);
    acted = delivered > 0 || acted;
    if (state->allocation_failed) {
      delivery_decision_free(&deliveries);
      set_error(error, ENGINE_ERROR_ALLOCATION, "contract delivery allocation failed");
      return false;
    }
  }
  delivery_decision_free(&deliveries);
  /* 14. */ ProcessingDecisionBuffer processing = {0};
  agent->choose_processing(agent, state, config, &processing);
  ENGINE_CHECK_ALLOC(processing, processing_decision_free, "processing decision");
  ENGINE_CHECK_STATE_ALLOC("processing decision");
  for (size_t i = 0; i < processing.count; i++) {
    const RecipeDef *recipe =
        config_find_recipe(config, processing.data[i].recipe_id);
    if (recipe != NULL)
      acted = processing_start_job(state, recipe, processing.data[i].batches,
                                   state->processing_capacity) ||
              acted;
    if (state->allocation_failed) {
      processing_decision_free(&processing);
      set_error(error, ENGINE_ERROR_ALLOCATION, "processing allocation failed");
      return false;
    }
  }
  processing_decision_free(&processing);
  /* 15. */ SalesDecisionBuffer sales = {0};
  agent->choose_sales(agent, state, config, &sales);
  ENGINE_CHECK_ALLOC(sales, sales_decision_free, "sales decision");
  ENGINE_CHECK_STATE_ALLOC("sales decision");
  for (size_t i = 0; i < sales.count; i++) {
    const ChannelDef *channel =
        config_find_channel(config, sales.data[i].channel_id);
    if (channel == NULL)
      continue;
    int sold = 0;
    bool exact = sales.data[i].quality != SALE_QUALITY_ANY;
    (void)markets_sell(state, sales.data[i].item_id, sales.data[i].quantity,
                       channel, exact, sales.data[i].quality, false,
                       QUALITY_REJECTED, &sold);
    acted = sold > 0 || acted;
    if (state->allocation_failed) {
      sales_decision_free(&sales);
      set_error(error, ENGINE_ERROR_ALLOCATION, "sale allocation failed");
      return false;
    }
  }
  sales_decision_free(&sales);
  /* 16. */
  for (size_t i = 0; i < config->upgrade_count; i++) {
    const UpgradeDef *upgrade = &config->upgrades[i];
    if (!state->upgrades_owned[upgrade->id]) {
      bool should_buy = agent->should_buy_upgrade(agent, state, upgrade->id);
      ENGINE_CHECK_STATE_ALLOC("upgrade decision");
      if (should_buy) {
        acted = actions_buy_upgrade(state, upgrade) || acted;
        ENGINE_CHECK_STATE_ALLOC("upgrade");
      }
    }
  }
  ENGINE_CHECK_STATE_ALLOC("upgrade");
  /* 17. Water and fertilize in ONE pass over the planted crops, per crop --
   * simulation/engine.py:176-183 is a single `for planted in
   * list(player.planted)` loop doing both. Splitting it into a water pass
   * followed by a fertilize pass looks equivalent and is not: both actions
   * spend money, so the split reorders the debits (W,W,F instead of W,F,W)
   * and every later balance rounds differently, which showed up as 1-256 ulp
   * drift in final_money/lowest_money against Python. It also changes
   * behaviour outright whenever cash is tight, since each action's
   * affordability check (`money < cost`) then sees a different balance.
   * RNG order is unaffected either way -- rng_chance below is the only draw
   * in this phase and its per-crop order is the same -- which is exactly why
   * the split survived the fixture suites undetected. */
  for (size_t i = 0; i < state->planted.count; i++) {
    PlantedCrop *planted = &state->planted.data[i];
    const CropDef *crop = config_find_crop(config, planted->crop_item_id);
    if (crop == NULL)
      continue;
    bool should_water = agent->should_water(agent, state, (int)i);
    ENGINE_CHECK_STATE_ALLOC("water decision");
    if (should_water && rng_chance(rng, agent->watering_diligence)) {
      acted = actions_water_crop(state, planted, &config->watering) || acted;
    }
    /* Operand order matches Python's `should_fertilize(...) and not
     * planted.fertilized`; no agent's should_fertilize has side effects, so
     * this is short-circuit-equivalent, but keep it in the reference's
     * order. */
    bool should_fertilize = agent->should_fertilize(agent, state, (int)i);
    ENGINE_CHECK_STATE_ALLOC("fertilizer decision");
    if (should_fertilize && !planted->fertilized) {
      if (state->fertilizer_inventory == 0)
        (void)actions_buy_fertilizer(state, &config->fertilizer, 1);
      acted =
          actions_fertilize_crop(state, planted, &config->fertilizer) || acted;
    }
  }
  ENGINE_CHECK_STATE_ALLOC("crop care");
  /* 18. */ acted = plant_open_slots(state, agent) || acted;
  ENGINE_CHECK_STATE_ALLOC("planting");
  /* 19. */ contracts_resolve_expired(state, config);
  /* 20. */ (void)inventory_collect_storage_liability(state, liability);
  /* 21. */ if (!acted)
    actions_do_nothing(state);
  /* 22. */ if (state->money < state->lowest_money)
    state->lowest_money = state->money;
  farm_state_track_peak_cash(state);
  if (no_viable_reinvestment(state)) {
    state->bankrupt = true;
    state->bankruptcy_day = state->day + 1;
    free(state->bankruptcy_reason);
    state->bankruptcy_reason = malloc(sizeof("no_viable_reinvestment"));
    if (state->bankruptcy_reason == NULL) {
      farm_state_mark_allocation_failed(state);
      set_error(error, ENGINE_ERROR_ALLOCATION, "bankruptcy reason allocation failed");
      return false;
    }
    strcpy(state->bankruptcy_reason, "no_viable_reinvestment");
  }
  /* 23. */ state->day++;
  if (callback != NULL)
    callback(state, &weather, context);
  return true;
}

bool engine_run_day(FarmState *state, const Agent *agent, FarmRng *rng,
                    EngineError *error) {
  return engine_run_day_observed(state, agent, rng, NULL, NULL, error);
}
